core/threaded/scheduler_NP.c

```diff
@@ -0,0 +1,403 @@
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***************/
+
+/** 
+ * Non-preemptive scheduler for the threaded runtime of the C target of Lingua
+ * Franca.
+ *  
+ * @author{Soroush Bateni <soroush@utdallas.edu>}
+ * @author{Edward A. Lee <eal@berkeley.edu>}
+ * @author{Marten Lohstroh <marten@berkeley.edu>}
+ */
+
+#ifndef NUMBER_OF_WORKERS
+#define NUMBER_OF_WORKERS 1
+#endif // NUMBER_OF_WORKERS
+
+#include "scheduler.h"
+#include "../platform.h"
+#include "../utils/semaphore.c"
+#include "../utils/vector.c"
+#include "scheduler_sync_tag_advance.c"
+#include <assert.h>
+
+#ifndef MAX_REACTION_LEVEL
+#define MAX_REACTION_LEVEL INITIAL_REACT_QUEUE_SIZE
+#endif
+
+/////////////////// External Variables /////////////////////////
+extern lf_mutex_t mutex;
+
+/////////////////// Scheduler Variables and Structs /////////////////////////
+/**
+ * @brief Atomically keep track of how many worker threads are idle.
+ *
+ * Initially assumed that there are 0 idle threads.
+ */
+semaphore_t* _lf_sched_semaphore; 
+
+/**
+ * @brief Indicate whether the program should stop
+ * 
+ */
+volatile bool _lf_sched_should_stop = false;
+
+/**
+ * @brief Array of reaction queues.
+ * 
+ * Each element is a reaction queue for a reaction level.
+ * 
+ */
+vector_t _lf_sched_vector_of_reaction_qs[MAX_REACTION_LEVEL + 1];
+
+/**
+ * @brief Mutexes for the reaction queues.
+ * 
+ * In case all threads are idle, there is no need to lock any of these mutexes.
+ * Otherwise, depending on the situation, the appropriate mutex for a given
+ * level must be locked before accessing the reaction queue of that level.
+ * 
+ */
+lf_mutex_t _lf_sched_vector_of_reaction_qs_mutexes[MAX_REACTION_LEVEL + 1];
+
+/**
+ * @brief Mutexes for the reaction queues.
+ * 
+ * In case all threads are idle, there is no need to lock any of these mutexes.
+ * Otherwise, depending on the situation, the appropriate mutex for a given
+ * level might need to be locked before accessing the reaction queue of that level.
+ * 
+ */
+volatile int _lf_sched_level_indexes[MAX_REACTION_LEVEL + 1] = {0};
```

Comment:
I dabbled with an atomic vector implementation for a bit, but in the end found it to need a lock when it is expanding, so I gave up on it. If you would like to take a stab at it, I might have a half-finished version somewhere.

Reply:
Oh, that is a good point -- it sounds like it would have to either never resize or rely on a lock instead of atomic operations. I would be happy to take that as an action item, but my instinct is that it is not a good idea to let that get in the way of this PR.

Reply:
There is an optimization that could make the vector access lock-free, at least for this scheduler. If we knew beforehand the maximum number of reactions per level, we could just statically pre-allocate the necessary memory. After talking to @edwardalee about it, we concluded that it would make the most sense to do this after the scalability-banks branch is merged in.